scheds/rust/scx_rusty/src/bpf/main.bpf.c

```diff
@@ -605,6 +612,41 @@ s32 BPF_STRUCT_OPS(rusty_select_cpu, struct task_struct *p, s32 prev_cpu,
 			}
 		}
 	}
+	
+	/*
+	 * The task might also be forked/execd so check for wake up only
+	 */
+	if (wake_flags & SCX_WAKE_TTWU) {
+		u64 highest_vtime = 0;
+		s32 found_cpu = -1;
+		u64 current_vtime = p->scx.dsq_vtime;
+		struct cpu_task_data *ctd;
+		/*
+			* Look for the currently running task with the highest vtime
+			* and with a vtime higher than the current task
+			*/
+		bpf_for(cpu, 0, nr_cpus) {
+			if (cpu_to_dom_id(cpu) != taskc->dom_id) {
+				continue;
+			}
+			ctd = MEMBER_VPTR(cpu_task_data, [cpu]);
+			if (!ctd) {
+				continue;
+			}
+			if (!ctd->has_running_task) {
+				found_cpu = cpu;
+				break;
+			} else if (vtime_before(current_vtime, ctd->dsq_vtime) && vtime_before(highest_vtime, ctd->dsq_vtime)) {
+				highest_vtime = ctd->dsq_vtime;
+				found_cpu = cpu;
+			}
+		}
+		if (found_cpu >= 0) {
+			cpu = found_cpu;
+			scx_bpf_kick_cpu(found_cpu, SCX_KICK_PREEMPT);
+			goto direct;
+		}
+	}
```

Comment:
It would be nice if we could somehow avoid having to do this O(n) iteration -- this won't work well for even moderately large systems. A couple of thoughts come to mind:

1. Maybe we could just have a max vruntime global in the domc, along with the cpu that has the task executing it? It would be racy so maybe not super ideal, though we could also use a spinlock. it's too bad vruntime isn't a u32 or we could have put that and the cpu into a single u64 and used CAS to update them both atomically, though that's also not without cost.
2. Perhaps we could have a BPF rbtree in the dom that was sorted according to max vruntime of the tasks currently running on a core? That way the lookup would just be O(log(n))

Reply:
So, if we keep track of the max start vtime of a domain, we can make the decision to preempt or not quickly. Once we decide to preempt, we'll need to scan to either update the new maximum. If we can make the iteration quicker by making it iterate only the CPUs in each domain rather than iterating all CPUs, this should be fine in most cases given that each domain doesn't tend to be too big. I'd like to avoid keeping per-domain rbtree if possible. That likely is more expensive in most cases.

Reply:
> If we can make the iteration quicker by making it iterate only the CPUs in each domain rather than iterating all CPUs

Are the CPU ids contiguous within each domain? Or do we have to keep an array of CPU ids for each domain?

> Once we decide to preempt, we'll need to scan to either update the new maximum.

Shouldn't we only wait to do this until the task actually starts running on a CPU?

> Maybe we could just have a max vruntime global in the domc, along with the cpu that has the task executing it? It would be racy so maybe not super ideal, though we could also use a spinlock.

Correct me if I'm wrong but this part of the algo feels like best effort anyway so maybe it's ok if sometimes we don't preempt the maximum vtime task but get close enough and at least ensure that the vtime is less for the preemptor.

Reply:
> > If we can make the iteration quicker by making it iterate only the CPUs in each domain rather than iterating all CPUs
> 
> Are the CPU ids contiguous within each domain? Or do we have to keep an array of CPU ids for each domain?

They aren't, so we'll have to introduce something to help iteration.

> > Once we decide to preempt, we'll need to scan to either update the new maximum.
> 
> Shouldn't we only wait to do this until the task actually starts running on a CPU?

Hmmm... I'm not sure. If we decide to preempt a CPU and on our way to send the preempting task to that CPU, wouldn't it make sense to consider that CPU to be occupied by the new task from then on? Otherwise, we'd be more prone to making multiple preemption decisions that target the same CPU.

> > Maybe we could just have a max vruntime global in the domc, along with the cpu that has the task executing it? It would be racy so maybe not super ideal, though we could also use a spinlock.
> 
> Correct me if I'm wrong but this part of the algo feels like best effort anyway so maybe it's ok if sometimes we don't preempt the maximum vtime task but get close enough and at least ensure that the vtime is less for the preemptor.

Yeah, while not ideal, that may be okay. Note that it can become a cascading problem in principle. If you kick out a task off CPU and that task has a vtime early enough to kick another CPU in the domain, we can be in an awkward situation.